File: 0806. Number of Lines To Write String/0806.java
// time complexity: O(n) where n = s.length()
// space complexity: O(1)
class Solution {
    // greedy scan with an imperative loop
    public int[] numberOfLines(int[] widths, String s) {
        int lines = 1;          // start on the first line
        int curWidth = 0;       // current line width
        for (char c : s.toCharArray()) {
            int w = widths[c - 'a'];           // width of this character
            if (curWidth + w > 100) {          // would overflow current line?
                lines++;                       // move to next line
                curWidth = w;                  // this char starts the new line
            } else {
                curWidth += w;                 // otherwise append to current line
            }
        }
        return new int[] { lines, curWidth };  // [total lines, last line width]
    }
}

File: 0806. Number of Lines To Write String/0806.py
# time complexity: O(n), where n = len(s)
# space complexity: O(1)
class Solution(object):
    def numberOfLines(self, widths, s):
        """
        :type widths: List[int]
        :type s: str
        :rtype: List[int]
        """
        lines = 1      # start with one line
        cur = 0        # current line width
        for ch in s:
            w = widths[ord(ch) - ord('a')]  # get pixel width of ch
            # if adding this char would exceed 100px, wrap to new line
            if cur + w > 100:
                lines += 1
                cur = w     # this char starts the new line
            else:
                cur += w    # otherwise append to current line
        return [lines, cur]